### PR TITLE
Add side effect check workaround (#4994)

### LINKED ISSUE #4994: Is it possible to get rid of side effects when evaluating `api.tx` etc?
* **I'm submitting a**
  - Other

* **What is the current behavior and expected behavior?**

Tab completion is not working in some circumstances. 

Pressing tab(s) on the input `api.tx.` shows nothing:
```
$ deno repl --eval-file=https://deno.land/x/subshell@0.0.2/tini.ts
api is initializing. Please hold on...
Deno 1.23.1
exit using ctrl+d or close()
> api.tx.
```
Same is true for `api.query.`, `api.consts.`, `api.rpc.`, etc.

Expected tab completion to work without patching Deno

```
> api.tx.
system                      session                     claims                      electionProviderMultiPhase  ump                         name                        __defineSetter__
scheduler                   grandpa                     vesting                     voterList                   hrmp                        arguments                   hasOwnProperty
preimage                    imOnline                    utility                     configuration               parasDisputes               caller                      __lookupGetter__
babe                        democracy                   identity                    parasShared                 registrar                   constructor                 __lookupSetter__
timestamp                   council                     proxy                       paraInclusion               slots                       apply                       isPrototypeOf
indices                     technicalCommittee          multisig                    paraInherent                auctions                    bind                        propertyIsEnumerable
balances                    phragmenElection            bounties                    paras                       crowdloan                   call                        valueOf
authorship                  technicalMembership         childBounties               initializer                 xcmPallet                   toString                    toLocaleString
staking
```

* **What is the motivation for changing the behavior?**

To improve the dev experience of using https://deno.land/x/polkadot in Deno repl.

* **Please tell us about your environment:**

  - Version: deno.land/x/polkadot@0.0.1
  - Environment:
    - Other: Deno 1.23.1

  - Language:
    - TypeScript

Related: https://github.com/denoland/deno/issues/14967

Adding `--v8-flags=--trace-side-effect-free-debug-evaluate` shows that evaluating `api.tx` failed side effect check.
```
$ deno repl --v8-flags=--trace-side-effect-free-debug-evaluate --eval-file=https://deno.land/x/subshell@0.0.2/tini.ts
api is initializing. Please hold on...
Deno 1.23.1
exit using ctrl+d or close()
> api.tx.[debug-evaluate] Checking function  for side effect.
0x11230843cac1 <BytecodeArray[10]>[debug-evaluate] Checking function get tx for side effect.
0x11230843cb95 <BytecodeArray[13]>[debug-evaluate] Checking function assertResult for side effect.
0x11230843cc79 <BytecodeArray[15]>[debug-evaluate] bytecode LdaModuleVariable may cause side effect.
[debug-evaluate] Function assertResult failed side effect check.
[debug-evaluate] Function assertResult failed side effect check.
```

Deno repl skips tab completion for expressions evaluated with side effects: https://github.com/denoland/deno/blob/8d82ba729937baf83011354242cabc3d50c13dc2/cli/tools/repl/editor.rs#L132

I searched for similar problems in Deno's issue history, and it seems no one encountered the same tab completion problem before. 

My question: why evaluating api.tx introduces side effect and is it possible to get rid of it?

I understand that deno.land/x/polkadot is experimental, and Deno is not currently a supported target, but still would like to play with it more. Feel free to close this if it's off topic. 

Thanks.

### PR DESCRIPTION
**Pull request by btwiuse**
Manually inline [assertReturn](https://github.com/polkadot-js/common/blob/7b8902ce3f129098fbb55ebd9c293d69fdc12ca4/packages/util/src/assert.ts#L39-L43) to pass v8 side effect check, which fixes tab completion when used in Deno.

Closes #4994.

### CODE REVIEW

File: packages/api/src/base/Getters.ts [hunk 1]
```diff
@@ -21,7 +21,10 @@ interface PkgJson {
 }
 
 function assertResult<T> (value: T | undefined): T {
-  return assertReturn(value, 'Api needs to be initialized before using, listen on \'ready\'');
+  if (isUndefined(value)) {
+    throw new Error('Api needs to be initialized before using, listen on \'ready\'');
+  }
+  return value;
```
Comment (jacogr) [inline packages/api/src/base/Getters.ts:27]:
```suggestion
  if (isUndefined(value)) {
    throw new Error('Api needs to be initialized before using, listen on \'ready\'');
  }
  
  return value;
```

I believe CI will complain here.

**Review by jacogr — COMMENTED** [review]
Perfect. Was wondering if this would work. (It is actually slightly faster as well, so double-bonus)

File: packages/api/src/base/Getters.ts [hunk 2]
```diff
@@ -21,7 +21,11 @@ interface PkgJson {
 }
 
 function assertResult<T> (value: T | undefined): T {
-  return assertReturn(value, 'Api needs to be initialized before using, listen on \'ready\'');
+  if (isUndefined(value)) {
+    throw new Error('Api needs to be initialized before using, listen on \'ready\'');
+  }
+
```
Comment (jacogr) [inline packages/api/src/base/Getters.ts:27]:
```suggestion

```

I added spaces :(

**Comment by btwiuse** [comment]
Fixed with `eslint --fix`. I think I need to do a little verification that it actually works. So don't merge it yet.

The master branch is broken for deno. I'm going to cherry-pick it to a stable branch and test it with a script.

**Comment by btwiuse** [comment]
After inlining all external dependencies of `assertResult`, I can verify that it is working. Steps:

```
git clone -b side-effect-check-workaround https://github.com/btwiuse/api $(mktemp -d)

cd $_

git checkout -b pr_5009_verification v8.10.1

git cherry-pick origin/master...origin/side-effect-check-workaround

yarn

yarn build

cat <<EOF >init.ts
import { ApiPromise, WsProvider } from './packages/api/build-deno/mod.ts';

async function initApi(){
  const PROVIDER = Deno.env.get("PROVIDER") ?? "wss://rpc.polkadot.io";
  const provider = new WsProvider( PROVIDER );
  return await ApiPromise.create({ provider });
}

console.log("api is initializing. Please hold on...");

const api = await initApi();
EOF

deno repl --v8-flags=--trace-side-effect-free-debug-evaluate --eval-file=init.ts
```

Result: tab completion for `api.tx` and friends is working:
```
api is initializing. Please hold on...
Deno 1.23.1
exit using ctrl+d or close()

> api.tx.[debug-evaluate] Checking function  for side effect.
0x25b207953cad <BytecodeArray[10]>[debug-evaluate] Checking function get tx for side effect.
0x25b207953d81 <BytecodeArray[13]>[debug-evaluate] Checking function assertResult for side effect.

system                      identity                    crowdloan
scheduler                   proxy                       xcmPallet
preimage                    multisig                    length
babe                        bounties                    name
timestamp                   childBounties               arguments
indices                     tips                        caller
balances                    electionProviderMultiPhase  constructor
authorship                  voterList                   apply
staking                     configuration               bind
session                     parasShared                 call
grandpa                     paraInclusion               toString
imOnline                    paraInherent                __defineGetter__
democracy                   paras                       __defineSetter__
council                     initializer                 hasOwnProperty
technicalCommittee          dmp                         __lookupGetter__
phragmenElection            ump                         __lookupSetter__
technicalMembership         hrmp                        isPrototypeOf
treasury                    parasDisputes               propertyIsEnumerable
claims                      registrar                   valueOf
vesting                     slots                       toLocaleString
utility                     auctions
> api.rpc.0x25b204e0dc75 <BytecodeArray[13]>[debug-evaluate] Checking function  for side effect.
0x25b204e19fc9 <BytecodeArray[10]>[debug-evaluate] Checking function get rpc for side effect.

author                net                   state                 __lookupSetter__
babe                  web3                  syncstate             isPrototypeOf
beefy                 eth                   system                propertyIsEnumerable
chain                 grandpa               constructor           toString
childstate            mmr                   __defineGetter__      valueOf
contracts             offchain              __defineSetter__      toLocaleString
dev                   payment               hasOwnProperty
engine                rpc                   __lookupGetter__

> api.query.[debug-evaluate] Checking function  for side effect.
0x2558078fa981 <BytecodeArray[10]>[debug-evaluate] Checking function get query for side effect.
0x2558078faa59 <BytecodeArray[13]>[debug-evaluate] Checking function assertResult for side effect.

substrate                   treasury                    hrmp
system                      claims                      paraSessionInfo
scheduler                   vesting                     parasDisputes
preimage                    identity                    registrar
babe                        proxy                       slots
timestamp                   multisig                    auctions
indices                     bounties                    crowdloan
balances                    childBounties               xcmPallet
transactionPayment          tips                        constructor
authorship                  electionProviderMultiPhase  __defineGetter__
staking                     voterList                   __defineSetter__
offences                    configuration               hasOwnProperty
session                     parasShared                 __lookupGetter__
grandpa                     paraInclusion               __lookupSetter__
imOnline                    paraInherent                isPrototypeOf
democracy                   paraScheduler               propertyIsEnumerable
council                     paras                       toString
technicalCommittee          initializer                 valueOf
phragmenElection            dmp                         toLocaleString
technicalMembership         ump

> api.consts.0x2558078fab3d <BytecodeArray[24]>[debug-evaluate] Checking function  for side effect.
0x2558078fb89d <BytecodeArray[10]>[debug-evaluate] Checking function get consts for side effect.

system                      claims                      auctions
scheduler                   vesting                     crowdloan
babe                        utility                     constructor
timestamp                   identity                    __defineGetter__
indices                     proxy                       __defineSetter__
balances                    multisig                    hasOwnProperty
transactionPayment          bounties                    __lookupGetter__
authorship                  childBounties               __lookupSetter__
staking                     tips                        isPrototypeOf
grandpa                     electionProviderMultiPhase  propertyIsEnumerable
imOnline                    voterList                   toString
democracy                   paras                       valueOf
phragmenElection            registrar                   toLocaleString
treasury                    slots

> api.runtimeChain.0x255805a7fc11 <BytecodeArray[5]>[debug-evaluate] Checking function  for side effect.
0x255804d921ed <BytecodeArray[10]>[debug-evaluate] Checking function get runtimeChain for side effect.

0                     toHuman               italics               sup
1                     toJSON                lastIndexOf           startsWith
2                     toRawType             link                  trim
3                     toString              localeCompare         trimStart
4                     toU8a                 match                 trimLeft
5                     anchor                matchAll              trimEnd
6                     at                    normalize             trimRight
7                     big                   padEnd                toLocaleLowerCase
createdAtHash         blink                 padStart              toLocaleUpperCase
initialU8aLength      bold                  repeat                toLowerCase
registry              charAt                replace               toUpperCase
length                charCodeAt            replaceAll            valueOf
constructor           codePointAt           search                __defineGetter__
encodedLength         concat                slice                 __defineSetter__
hash                  endsWith              small                 hasOwnProperty
isEmpty               fontcolor             split                 __lookupGetter__
eq                    fontsize              strike                __lookupSetter__
inspect               fixed                 sub                   isPrototypeOf
setOverride           includes              substr                propertyIsEnumerable
toHex                 indexOf               substring             toLocaleString

...
```

**Comment by jacogr** [comment]
What is broken for Deno on master? (Only thing that has been going in is the runtime stuff, which should just be extra decoration. Could be a mismatch between master and the deployed version if that is the cause)

**Review by jacogr — APPROVED** [review]
Thank you! Really happy this fixes the underlying issue for the specific use.

**Comment by btwiuse** [comment]
> What is broken for Deno on master? (Only thing that has been going in is the runtime stuff, which should just be extra decoration. Could be a mismatch between master and the deployed version if that is the cause)

```
git checkout origin/master

yarn

yarn build

cat <<EOF >init.ts
import { ApiPromise, WsProvider } from './packages/api/build-deno/mod.ts';

async function initApi(){
  const PROVIDER = Deno.env.get("PROVIDER") ?? "wss://rpc.polkadot.io";
  const provider = new WsProvider( PROVIDER );
  return await ApiPromise.create({ provider });
}

console.log("api is initializing. Please hold on...");

const api = await initApi();
EOF

deno repl --v8-flags=--trace-side-effect-free-debug-evaluate --eval-file=init.ts
```

Output:
```
$ deno repl --v8-flags=--trace-side-effect-free-debug-evaluate --eval-file=init.ts
Download https://deno.land/x/polkadot/api-base/types/calls.ts
Download https://deno.land/x/polkadot/api/types/calls.ts
Warning Implicitly using latest version (0.0.1) for https://deno.land/x/polkadot/api-base/types/calls.ts
Download https://deno.land/x/polkadot@0.0.1/api-base/types/calls.ts
Warning Implicitly using latest version (0.0.1) for https://deno.land/x/polkadot/api/types/calls.ts
Download https://deno.land/x/polkadot@0.0.1/api/types/calls.ts
error in --eval-file file init.ts. Uncaught SyntaxError: The requested module 'https://deno.land/x/polkadot/types-known/mod.ts' does not provide an export named 'getSpecRuntime'
    at async <anonymous>:2:38
Deno 1.23.1
exit using ctrl+d or close()
> api
Uncaught ReferenceError: api is not defined
    at <anonymous>:2:1
> 
```

Probably related to a recent change (somewhere after v8.10.1)

**Comment by jacogr** [comment]
Yes, that would make sense. This is new in the master version, but not published to deno as of yet. (So it tries to pull from the remote and cannot find it)

**Comment by jacogr** [comment]
Merged without waiting for GH to report back, it is 100% sane, just seems to be some CI hiccup along the way.

**Pull request merged**